### ISSUE TITLE
[vSphere] machine-controller running with CCM credentials

### ISSUE THREAD

**Issue by heilerich** [kubermatic/kubermatic#11980]
### What happened?

We are currently trying to minimise the access cluster administrators have on the vSphere infrastructure. Kubermatic already provides separate credentials for the cloud provider functionality and an 'infra mangement user' used by KKP. The credentials for the cloud provider functionality are exposed to user cluster administrators, because some of the components using the credentials are running inside the user cluster. The infra management credentials stay on the seed.

We now noticed that the `machine-controller` deployments are using the cloud provider credentials instead of the infra management credentials, despite running on the seed and apparently managing infrastructure, i.e. virtual machines. This forces us to give the credentials exposed to cluster administrators more permissions than we would like to.

Is this intentional?

### Expected behavior

The `machine-controller` and `machine-controller-webhook` deployments use the `infraManagementUser/Password` from the cloud-credentials secret instead of the `username/password` keys (if available).

### How is your environment configured?

- KKP version: v2.21.6
- Shared or separate master/seed clusters?: shared

**Comment by embik** [comment]
Hey @heilerich, the issue with the infra management credentials is that they would be the same for all user clusters on the seed since they are configured per datacenter, not per user cluster. We should look into improving the configuration key's documentation, but machine-controller is kind of considered part of the cloud provider functionality. Since vSphere is the only provider featuring this kind of setting, it would be very unexpected if machine-controller wasn't using the credentials passed to the cluster.

I understand where you're coming from though, reducing exposure of credentials to cluster admins is good practice. For some providers, we are already running CCMs within the seed cluster for that reason. We are also looking into moving more CSI components into the seed as per #11985. That will hopefully eliminate the need to sync credentials into the user cluster. 

Would that help you as well?

**Comment by heilerich** [comment]
Thank you for the clarification @embik.

Our ideal scenario would be to isolate cluster administrators as much as possible / completely from the underlying vSphere infrastructure.

At the moment we are prototyping a process that does the following
(1) Clusters are provisioned from a preset containing 'dummy' credentials
(2) A custom operator process watches new cluster resources, creates a dedicated minimal set of credentials on vSphere and patches them into the cluster resource.
(3) Cluster admins only have access to the dedicated cluster credentials with slightly restricted (but still to wide) permissions

Much better of course would be if we could keep the vSphere credentials out of the cluster all together!

Now that KKP does not put a cloud config override containing the credentials into the MachineDeployments anymore, the CSI driver would be the only component, that spills the credentials into the tenant clusters. That is, if I am not missing something critical here.

Long story short: The changes proposed in #11985 would be an absolute game changer for us since we wouldn't have to worry about creating dedicated credentials and imposing restrictions anymore!

Regarding this issue, it sounds like we can't do anything about the permissions required by machine-controller, so I'll close this for now. But I will keep a close eye on #11985 :)
